sfxclient/multitokensink.go

```diff
@@ -121,207 +139,331 @@ func newWorker(maxBuffer int, errorHandler func(error) error) *worker {
-func (w *datapointWorker) processMsg(msg *msg) (stop bool) {
-	if msg == stopMsg {
-		stop = true
+func (w *datapointWorker) handleError(err error, token string, datapoints []*datapoint.Datapoint) {
+	if err != nil {
+		_ = atomic.AddInt64(&w.stats.TotalDatapointsDropped, int64(len(datapoints)))
+		// TODO: improve httpSinks handle response to return the response code
+		// parse the http status code from the error returned by the httpsink
+		statusCode := "unknown"
+		if w.invalidStatusRegexp.MatchString(err.Error()) {
+			statusCode = w.statusCodeRegexp.FindStringSubmatch(err.Error())[0]
+		}
+		if counter, ok := w.stats.TotalDatapointsDroppedByToken.Load(token, statusCode); ok {
+			_ = atomic.AddInt64(counter, int64(len(datapoints)))
+		} else {
+			w.stats.TotalDatapointsDroppedByToken.Store(token, statusCode, int64(len(datapoints)))
+		}
+		_ = w.errorHandler(err)
 	} else {
-		var remainingBuffer int
-		var msgLength int
-		var data = msg.data.([]*datapoint.Datapoint) // cast msg data to an array of points (avoiding reflection)
-		// iterate over the data in the message until it's empty, breaking it up into batches
-		for len(data) > 0 {
-			remainingBuffer = (w.maxBuffer - len(w.buffer))
-			msgLength = len(data)
-			if msgLength >= w.maxBuffer && remainingBuffer >= w.maxBuffer {
-				w.emit(msg.token, data[:w.maxBuffer], nil)
-				data = data[w.maxBuffer:]
-			} else if msgLength > remainingBuffer {
-				w.buffer = append(w.buffer, data[:remainingBuffer]...)
-				data = data[remainingBuffer:]
-			} else if remainingBuffer >= msgLength {
-				w.buffer = append(w.buffer, data[:msgLength]...)
-				data = data[msgLength:]
-			}
-			if len(w.buffer) >= w.maxBuffer {
-				// if the buffer is full emit the data
-				w.emit(msg.token, w.buffer, nil)
-				w.buffer = w.buffer[:0] // reset the buffer
-			}
+		_ = atomic.AddInt64(&w.stats.TotalDatapointsEmitted, int64(len(datapoints)))
+	}
+}
+
+// processMsg is
+func (w *datapointWorker) processMsg(msg *dpMsg) {
+	for len(msg.data) > 0 {
+		msgLength := len(msg.data)
+		remainingBuffer := (w.batchSize - len(w.buffer))
+		if msgLength > remainingBuffer {
+			msgLength = remainingBuffer
+		}
+		w.buffer = append(w.buffer, msg.data[:msgLength]...)
+		msg.data = msg.data[msgLength:]
+		if len(w.buffer) >= w.batchSize {
+			w.emit(msg.token, w.buffer)
+			w.buffer = w.buffer[:0]
 		}
 	}
-	return
 }
 
 // bufferDatapoints is responsible for batching incoming datapoints into a buffer
-func (w *datapointWorker) bufferFunc(msg *msg) (stop bool) {
+func (w *datapointWorker) bufferFunc(msg *dpMsg) (stop bool) {
 	var lastTokenSeen = msg.token
-	_ = w.processMsg(msg)
+	w.processMsg(msg)
 outer:
-	for len(w.buffer) < w.maxBuffer {
+	for len(w.buffer) < w.batchSize {
 		select {
 		case msg = <-w.input:
 			if msg.token != lastTokenSeen {
 				// if the token changes, then emit what ever is in the buffer before proceeding
-				w.emit(msg.token, w.buffer, nil)
+				w.emit(msg.token, w.buffer)
```

Comment:
you want to emit with the lastTokenSeen no? that's what's in the buffer isn't it?

Reply:
Yes you are correct, good catch!